amm/contracts/stable_pool/lib.rs

```diff
@@ -0,0 +1,1033 @@
+                    let events = self.psp22.mint(fee_to, admin_fee)?;
+                    self.emit_events(events);
+                }
+            }
+
+            // update reserves
+            for (i, &amount) in amounts.iter().enumerate() {
+                self.increase_reserve(i, amount)?;
+            }
+
+            self.env().emit_event(AddLiquidity {
+                provider: self.env().caller(),
+                token_amounts: amounts,
+                shares,
+                to,
+            });
+            Ok((shares, fee_part))
+        }
+
+        // Note that this method does not require to update rates, neither it uses rates.
+        // Thus it's always possible to call it, even if the rate is outdated, or the rate provider is down.
+        #[ink(message)]
+        fn remove_liquidity_by_shares(
+            &mut self,
+            shares: u128,
+            min_amounts: Vec<u128>,
+            to: AccountId,
+        ) -> Result<Vec<u128>, StablePoolError> {
+            let amounts = math::compute_withdraw_amounts_for_lp(
+                shares,
+                &self.reserves(),
+                self.psp22.total_supply(),
+            )?;
+
+            // Check if enough tokens are withdrawn
+            ensure!(
+                amounts
+                    .iter()
+                    .zip(min_amounts.iter())
+                    .all(|(amount, min_amount)| amount >= min_amount),
+                StablePoolError::InsufficientOutputAmount
+            );
+
+            // transfer tokens
+            for (&token, &amount) in self.pool.tokens.iter().zip(amounts.iter()) {
+                self.token_by_address(token).transfer(to, amount, vec![])?;
+            }
+
+            // update reserves
+            for (i, &amount) in amounts.iter().enumerate() {
+                self.decrease_reserve(i, amount)?;
+            }
+
+            // Burn liquidity
+            let events = self.psp22.burn(self.env().caller(), shares)?;
+            self.emit_events(events);
+
+            Ok(amounts)
+        }
+
+        #[ink(message)]
+        fn remove_liquidity_by_amounts(
+            &mut self,
+            max_share_amount: u128,
+            amounts: Vec<u128>,
+            to: AccountId,
+        ) -> Result<(u128, u128), StablePoolError> {
+            // Make sure rates are up to date before we attempt any calculations
+            self.update_rates();
+            if amounts.len() != self.pool.tokens.len() {
+                return Err(StablePoolError::IncorrectAmountsCount);
+            }
+
+            let rates = self.get_scaled_rates()?;
+
+            // calc comparable amounts
+            let (shares_to_burn, fee_part) = math::rated_compute_lp_amount_for_withdraw(
+                &rates,
+                &amounts,
+                &self.reserves(),
+                self.psp22.total_supply(),
+                Some(&self.pool.fees),
+                self.amp_coef(),
+            )?;
```

Comment:
This is the same code as in [`get_burn_liquidity_for_amounts`](https://github.com/Cardinal-Cryptography/common-amm-stable-swap/pull/3/files#diff-a97a9d61ac214aed0351684f2397db3d88753207096caf51cdffaae6893014faR828) so we can just call that method here instead.

Reply:
[Reuse StablePoolView getters](https://github.com/Cardinal-Cryptography/common-amm-stable-swap/pull/5/commits/c0a1498d40e52542621db55524f555f06ae12c56)